python/loyalty_workflow.py

```diff
@@ -0,0 +1,202 @@
+
+    async def _process_invite_guest(self, guest_id: str) -> None:
+        if len(self.customer.guests) >= self.customer.tier.guests_allowed:
+            await workflow.execute_activity(
+                LoyaltyActivities.send_email,
+                "Sorry, you need to earn more points to invite more guests!",
+                start_to_close_timeout=timedelta(seconds=5),
+            )
+            return
+
+        if guest_id not in self.customer.guests:
+            self.customer.guests.append(guest_id)
+
+        previous = StatusTier.previous(self.customer.tier)
+        guest = Customer(
+            id=guest_id,
+            tier=previous,
+        )
+
+        started: bool = await workflow.execute_activity(
+            LoyaltyActivities.start_guest_workflow,
+            guest,
+            start_to_close_timeout=timedelta(seconds=5),
+        )
+
+        if started:
+            await workflow.execute_activity(
+                LoyaltyActivities.send_email,
+                "Congratulations! Your guest has been invited!",
+                start_to_close_timeout=timedelta(seconds=5),
+            )
+        else:
+            await workflow.execute_activity(
+                LoyaltyActivities.send_email,
+                "Sorry, your guest has already canceled their account.",
+                start_to_close_timeout=timedelta(seconds=5),
+            )
+
+    @workflow.signal(name=Signals.ENSURE_MINIMUM_STATUS)
+    async def ensure_minimum_status(self, min_status: StatusTier) -> None:
+        self._signal_queue.append((Signals.ENSURE_MINIMUM_STATUS, min_status))
+
+    async def _process_ensure_minimum_status(self, min_status: StatusTier) -> None:
+        if self.customer.tier.level < min_status.level:
+            self.customer.tier = min_status
+            self.customer.points = min_status.minimum_points
+
+            await workflow.execute_activity(
+                LoyaltyActivities.send_email,
+                "Congratulations! You've been promoted to '{}' status!".format(self.customer.tier.name),
+                start_to_close_timeout=timedelta(seconds=5)
+            )
+
+    @workflow.query(name=Queries.GET_STATUS)
+    def get_status(self) -> GetStatusResponse:
+        return GetStatusResponse(
+            points=self.customer.points,
+            account_active=self.customer.account_active,
+            status_level=self.customer.tier.level,
+            tier=self.customer.tier
+        )
+
+    @workflow.query(name=Queries.GET_GUESTS)
+    def get_guests(self) -> list[str]:
+        return list(self.customer.guests)
+
+    def validate_inputs(self) -> None:
+        # Note: I found that without this, I got a "'dict' object has no attribute 'tier'" error thrown.
+        if isinstance(self.customer, dict):
+            self.customer = Customer(**self.customer)
+        if isinstance(self.customer.tier, dict):
+            self.customer.tier = StatusTier(**self.customer.tier)
```

Comment:
@cretz Looking for advice/help: With the input arg to a workflow as a dataclass, I was consistently seeing it treated as a plain dict rather than the actual dataclass. Curious if this is normal/expected, or if I'm doing something wrong here.

Reply:
We use the parameters' type hints to know what type to deserialize into. So a dataclass becomes a JSON object on the wire and we deserialize into the dataclass type from the type hint. If this is not happening, I wonder if there is a bug or something with the implementation (maybe related to the default parameter, unsure).

Also, can you make sure you run a type checker on your code? Our type-safe client code usually expects only a single parameter (and we strongly encourage no more than a single dataclass parameter).

Reply:
I tested it without the default value on the second parameter, and the first arg does come in as the correct type. So, default value appears to be the culprit here.

I ran `mypy --strict .` on the project, and got nothing about the workflow method's typing. I *do* however get errors about `execute_activity`, for example:

```
loyalty_workflow.py:51: error: Argument 1 to "execute_activity" has incompatible type "Callable[[LoyaltyActivities, str], Coroutine[Any, Any, None]]"; expected "CallableAsyncSingleParam[str, <nothing>]"  [arg-type]
```

Looks like this is because I have the activities defined in a class (`LocalActivities`). I did this so that they can have access to the client in the worker. Is there a more correct/recommended way to do this?

In the meantime, I'll convert the workflow to use a single dataclass.

Reply:
> So, default value appears to be the culprit here.

I will file an issue on this. Right now we require the parameter count to be the same as the input count before using the type hints: https://github.com/temporalio/sdk-python/blob/393e5c8924ff34698d57311e84947f1bd3da4b05/temporalio/worker/_activity.py#L328-L335. Specifically "we only use arg type hints if they match the input count". But I can change that to "we only use arg type hints if the non-defaulted ones at least match the input count".

> I do however get errors about execute_activity, for example:

I expected that. We only support single-parameter callables in typesafe invocations. But it's unrelated to your original issue here.

Reply:
Opened https://github.com/temporalio/sdk-python/issues/360